File: backend/app/domain/schemas.py
"""
Pydantic v2 schemas -- the wire contract between FastAPI and the Next.js client.

These models are the *only* shapes that cross the HTTP boundary. ORM objects are
never returned directly, for three reasons:

1.  It stops internal columns leaking into a public response by accident.
2.  It gives FastAPI a real OpenAPI schema, which is what makes the mirrored
    TypeScript types in ``frontend/src/lib/types.ts`` trustworthy.
3.  It converts paise to rupees in exactly one place, so no component in the UI
    ever has to remember the unit.

Naming convention: ``*Out`` = server -> client, ``*In`` = client -> server.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.domain.enums import (
    ActorType,
    AgentMode,
    AuditEventType,
    FailureCategory,
    GatewayMode,
    GuardrailDecision,
    PaymentMethod,
    PaymentStatus,
    RecoveryStatus,
    RecoveryStrategy,
    ToolCapability,
)

# ---------------------------------------------------------------------------
# Shared base
# ---------------------------------------------------------------------------


class ORMModel(BaseModel):
    """Base for schemas built straight from SQLAlchemy rows."""

    model_config = ConfigDict(from_attributes=True)


def paise_to_rupees(paise: int) -> float:
    """
    Presentation-only conversion.

    Rupees appear at the API edge and nowhere else. All arithmetic, comparison
    and storage upstream of this call is integer paise.
    """
    return round(paise / 100, 2)


# ---------------------------------------------------------------------------
# Customers & payments
# ---------------------------------------------------------------------------


class CustomerOut(ORMModel):
    id: str
    name: str
    email: str
    phone: str = ""
    risk_flagged: bool = False
    total_payments: int = 0
    successful_payments: int = 0
    prior_success_rate: float = 0.5
    lifetime_value_paise: int = 0
    lifetime_value_rupees: float = 0.0


class PaymentOut(ORMModel):
    id: str
    customer_id: str
    customer: CustomerOut | None = None

    amount_paise: int
    amount_rupees: float
    currency: str = "INR"
    method: PaymentMethod
    status: PaymentStatus
    description: str = ""

    razorpay_order_id: str | None = None
    razorpay_payment_id: str | None = None

    error_code: str | None = None
    error_source: str | None = None
    error_step: str | None = None
    error_reason: str | None = None
    error_description: str | None = None

    is_recovery_attempt: bool = False
    parent_payment_id: str | None = None

    #: Present on failed payments only: the id of the case opened for it, or
    #: ``None`` when the failure has not been analysed yet. Lets the payments
    #: table render "Analyse" vs "View case" without a second round trip.
    recovery_case_id: str | None = None

    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Guardrails
# ---------------------------------------------------------------------------


class GuardrailEvaluationOut(BaseModel):
    """
    Result of one guardrail rule.

    ``observed`` and ``limit`` are carried as display strings so the UI can show
    "3 attempts vs limit 2" without knowing anything about the rule's internals.
    That keeps new rules from requiring frontend changes.
    """

    rule_id: str = Field(examples=["R1_MAX_ATTEMPTS"])
    name: str = Field(examples=["Maximum recovery attempts"])
    description: str
    decision: GuardrailDecision
    passed: bool
    reason: str
    observed: str | None = None
    limit: str | None = None
    #: False when the strategy creates no payment attempt, so the rule had nothing
    #: to constrain. Distinguishes "checked and cleared" from "never consulted",
    #: which are otherwise identical on the wire.
    applicable: bool = True


class GuardrailVerdictOut(BaseModel):
    """Aggregate of every rule: the most restrictive decision wins."""

    decision: GuardrailDecision
    evaluations: list[GuardrailEvaluationOut]
    blocking_rules: list[str] = Field(default_factory=list, description="rule_ids that returned DENY")
    approval_rules: list[str] = Field(default_factory=list, description="rule_ids that forced human approval")
    summary: str


# ---------------------------------------------------------------------------
# Agent
# ---------------------------------------------------------------------------


class AgentToolCallOut(ORMModel):
    """One step of the reasoning trace, for the explainability panel."""

    id: str
    run_id: str
    step: int
    tool_name: str
    capability: ToolCapability
    arguments: dict[str, Any] = Field(default_factory=dict)
    result: dict[str, Any] = Field(default_factory=dict)
    ok: bool = True
    error: str | None = None
    latency_ms: int = 0
    created_at: datetime


class AgentRecoveryPlan(BaseModel):
    """
    The structured object the agent must produce -- LLM or rule-based alike.

    Two properties make this safe to accept from a language model:

    *   Every field is an enum or a bounded scalar. There is no free-form field
        that can express an amount, an account or an instruction to the gateway.
    *   ``model_config`` forbids extra keys, so a model that invents
        ``"amount_override": 100`` fails validation instead of being ignored.

    Note the absence of any money field. The recovery amount is copied from the
    original payment by the service layer; the agent is structurally incapable of
    influencing it.
    """

    model_config = ConfigDict(extra="forbid")

    failure_category: FailureCategory
    confidence: float = Field(ge=0.0, le=1.0)
    strategy: RecoveryStrategy
    #: Why, in the operator's language. Shown verbatim in the approval UI, so a
    #: human is judging the agent's actual reasoning rather than a summary of it.
    rationale: str = Field(min_length=10, max_length=1200)
    #: What the customer would be told. Kept separate from the rationale because
    #: the two audiences need different tone and different detail.
    customer_message: str = Field(min_length=10, max_length=500)
    #: Optional signals the model wants to record (e.g. "error_code=BAD_REQUEST").
    evidence: list[str] = Field(default_factory=list, max_length=8)

    @field_validator("rationale", "customer_message")
    @classmethod
    def _strip(cls, v: str) -> str:
        return v.strip()


class AgentRunOut(BaseModel):
    """Metadata about one analysis run, independent of its conclusions."""

    run_id: str
    mode: AgentMode
    model: str | None = None
    steps: int
    total_latency_ms: int
    degraded_reason: str | None = Field(
        default=None,
        description="Set when the LLM path was unavailable and the rule engine was used instead.",
    )


# ---------------------------------------------------------------------------
# ML propensity
# ---------------------------------------------------------------------------


class PropensityResultOut(BaseModel):
    """Prediction from the recovery-propensity model."""

    score: float = Field(ge=0.0, le=1.0, description="P(recovery attempt succeeds)")
    model_version: str
    #: Ordered, human-readable drivers behind this particular score.
    top_factors: list[str] = Field(default_factory=list)
    #: True when the trained artefact was missing and the documented heuristic
    #: fallback produced the score. Surfaced so nobody mistakes it for a model.
    is_fallback: bool = False


# ---------------------------------------------------------------------------
# Recovery cases
# ---------------------------------------------------------------------------


class RecoveryAttemptOut(ORMModel):
    id: str
    attempt_number: int
    strategy: RecoveryStrategy
    amount_paise: int
    amount_rupees: float
    #: Typed, like every other status on the wire. An attempt only ever holds the
    #: subset {executing, awaiting_payment, recovered, failed, expired}, all of
    #: which are valid members -- and typing it means a garbage value in the
    #: column fails at the boundary instead of serialising into the UI unnoticed.
    status: RecoveryStatus
    gateway_mode: GatewayMode
    razorpay_order_id: str | None = None
    razorpay_payment_id: str | None = None
    failure_reason: str | None = None
    created_at: datetime
    completed_at: datetime | None = None


class RecoveryCaseSummaryOut(ORMModel):
    """Row shape for list views. Deliberately light -- no trace, no evaluations."""

    id: str
    original_payment_id: str
    customer_id: str
    customer_name: str = ""
    status: RecoveryStatus
    failure_category: FailureCategory
    strategy: RecoveryStrategy
    guardrail_decision: GuardrailDecision
    propensity_score: float
    agent_mode: AgentMode
    amount_paise: int
    amount_rupees: float
    attempt_count: int
    created_at: datetime
    updated_at: datetime


class RecoveryCaseOut(RecoveryCaseSummaryOut):
    """Full case detail: everything the approval screen needs in one response."""

    classification_confidence: float
    agent_rationale: str
    customer_message: str
    propensity_model_version: str = ""
    propensity_is_fallback: bool = False

    guardrail_evaluations: list[GuardrailEvaluationOut] = Field(default_factory=list)
    policy_snapshot: dict[str, Any] = Field(default_factory=dict)

    approved_by: str | None = None
    approved_at: datetime | None = None
    rejected_by: str | None = None
    rejected_at: datetime | None = None
    rejection_reason: str | None = None

    recovered_at: datetime | None = None
    recovered_amount_paise: int = 0
    recovered_amount_rupees: float = 0.0
    failure_note: str | None = None
    expires_at: datetime | None = None

    original_payment: PaymentOut | None = None
    customer: CustomerOut | None = None
    attempts: list[RecoveryAttemptOut] = Field(default_factory=list)

    #: Precomputed by the service so the UI never re-implements policy logic.
    #: If the button's enabled state were derived in React, the frontend would
    #: become a second, divergent copy of the guardrail engine.
    can_approve: bool = False
    can_reject: bool = False
    approval_blocked_reason: str | None = None


class CheckoutSessionOut(BaseModel):
    """Everything the customer-facing recovery page needs to open Checkout."""

    case_id: str
    attempt_id: str
    order_id: str
    amount_paise: int
    amount_rupees: float
    currency: str
    #: Razorpay publishable key, or ``None`` in simulated mode. It is safe to
    #: expose: the key id is public by design; the secret never leaves the server.
    razorpay_key_id: str | None = None
    gateway_mode: GatewayMode
    customer_name: str
    customer_email: str
    customer_phone: str
    description: str
    expires_at: datetime | None = None


# ---------------------------------------------------------------------------
# Audit
# ---------------------------------------------------------------------------


class AuditEventOut(ORMModel):
    id: str
    sequence: int
    case_id: str | None = None
    payment_id: str | None = None
    event_type: AuditEventType
    actor_type: ActorType
    actor_id: str
    summary: str
    payload: dict[str, Any] = Field(default_factory=dict)
    prev_hash: str
    hash: str
    created_at: datetime


class AuditChainVerificationOut(BaseModel):
    """
    Result of recomputing the ledger's hash chain from genesis.

    A demo that only *claims* immutability proves nothing. This endpoint
    recomputes every hash and names the first sequence number that fails, which
    is what makes the guarantee inspectable by a reviewer.
    """

    valid: bool
    events_checked: int
    head_hash: str | None = None
    broken_at_sequence: int | None = None
    message: str


# ---------------------------------------------------------------------------
# Metrics & policy
# ---------------------------------------------------------------------------


class DashboardMetricsOut(BaseModel):
    """Merchant-level numbers for the landing dashboard."""

    total_volume_paise: int
    captured_volume_paise: int
    failed_volume_paise: int
    recoverable_volume_paise: int = Field(description="Failed volume that passed guardrails as recoverable")
    recovered_volume_paise: int

    total_volume_rupees: float
    captured_volume_rupees: float
    failed_volume_rupees: float
    recoverable_volume_rupees: float
    recovered_volume_rupees: float

    total_payments: int
    failed_payments: int
    unanalysed_failures: int

    cases_total: int
    cases_awaiting_approval: int
    cases_blocked: int
    cases_recovered: int
    cases_failed: int

    #: recovered_volume / recoverable_volume, as a percentage. 0.0 when there is
    #: nothing recoverable -- reported rather than hidden, so an empty database
    #: reads as "0% of nothing" instead of a crash or a misleading blank.
    recovery_rate_pct: float
    failure_rate_pct: float

    #: Value of recovery orders created today, against the daily budget cap.
    daily_budget_used_paise: int
    daily_budget_limit_paise: int


class FailureBreakdownItem(BaseModel):
    category: FailureCategory
    count: int
    volume_paise: int
    volume_rupees: float
    recovered_count: int


class PolicyOut(BaseModel):
    """
    The active guardrail configuration, exposed read-only.

    Read-only on purpose. If policy could be edited through the same API the
    agent's flow uses, the guardrails would be reachable from the automated path,
    and a limit an automated system can raise is not a limit.
    """

    max_recovery_attempts: int
    recovery_cooldown_seconds: int
    high_value_review_threshold_paise: int
    max_recovery_amount_paise: int
    daily_recovery_budget_paise: int
    max_cases_per_customer_per_day: int
    min_propensity_score: float
    max_payment_age_hours: int
    require_human_approval: bool
    auto_approve_enabled: bool
    auto_approve_max_paise: int
    auto_approve_min_propensity: float
    recovery_link_ttl_minutes: int
    non_recoverable_categories: list[FailureCategory]
    rules: list[dict[str, str]] = Field(description="Static catalogue: rule_id, name, description")


class SystemStatusOut(BaseModel):
    """Honest self-report of which subsystems are live vs simulated."""

    app: str
    version: str
    environment: str
    agent_mode: AgentMode
    gemini_model: str | None = None
    gateway_mode: GatewayMode
    ml_model_loaded: bool
    ml_model_version: str | None = None
    database: str
    #: Only conditions the typed fields above cannot express -- an interpreter
    #: outside the tested range, or a Gemini key that is present but unusable.
    #: Degraded modes themselves are read from `agent_mode`, `gateway_mode` and
    #: `ml_model_loaded`, so that no consumer has to render the same fact twice.
    warnings: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Request bodies
# ---------------------------------------------------------------------------


class SimulateFailureIn(BaseModel):
    """
    Demo helper: manufacture a realistic failed payment.

    Exists because a submission reviewer cannot make a real card decline on
    demand. Every field is optional; omitted fields are drawn from the seeded
    scenario catalogue so a single empty POST produces a plausible failure.
    """

    model_config = ConfigDict(extra="forbid")

    customer_id: str | None = None
    amount_paise: int | None = Field(default=None, ge=100, le=100_000_000)
    method: PaymentMethod | None = None
    #: Named scenario from ``app/db/seed.py`` (e.g. ``"bank_decline_card"``).
    scenario: str | None = None
    description: str | None = Field(default=None, max_length=255)


class FailureScenarioOut(BaseModel):
    """
    One entry from the demo failure catalogue.

    Served so the "simulate a failed payment" picker can be built from the
    catalogue the server actually holds. The alternative -- a hard-coded list in
    the React component -- was tried and drifted immediately: six of its eight
    ids named scenarios that had never existed, so the picker's most important
    option (the fraud case the guardrails must refuse) returned a 404. A list of
    server-side keys maintained by hand on the client is a list that will be
    wrong.
    """

    key: str = Field(description="Pass this as `scenario` to POST /api/payments/simulate-failure")
    label: str
    method: PaymentMethod
    expected_category: FailureCategory
    error_reason: str
    error_description: str


class AnalyzeIn(BaseModel):
    model_config = ConfigDict(extra="forbid")
    #: Force the deterministic planner even when a Gemini key is configured.
    #: Used by tests and to demonstrate the degraded path on stage.
    force_rule_based: bool = False


class ApproveIn(BaseModel):
    model_config = ConfigDict(extra="forbid")
    approved_by: str = Field(min_length=1, max_length=120, description="Operator identity, for the audit trail")
    note: str | None = Field(default=None, max_length=500)


class RejectIn(BaseModel):
    model_config = ConfigDict(extra="forbid")
    rejected_by: str = Field(min_length=1, max_length=120)
    reason: str = Field(min_length=1, max_length=500)


class VerifyPaymentIn(BaseModel):
    """
    Payload posted back by Razorpay Checkout on success.

    The signature is verified server-side with HMAC-SHA256 before anything is
    marked recovered. A client-side "payment succeeded" callback is a claim, not
    proof -- trusting it would let anyone mark any case recovered with curl.
    """

    model_config = ConfigDict(extra="forbid")

    razorpay_order_id: str = Field(min_length=1, max_length=64)
    razorpay_payment_id: str = Field(min_length=1, max_length=64)
    razorpay_signature: str = Field(min_length=1, max_length=256)


class MarkAttemptFailedIn(BaseModel):
    """Demo helper: force the current attempt to fail, to exercise the failure path."""

    model_config = ConfigDict(extra="forbid")
    reason: str = Field(default="Customer did not complete the payment", max_length=500)


class ErrorOut(BaseModel):
    """Uniform error envelope for every non-2xx response."""

    error: str = Field(description="Stable machine-readable code, e.g. 'guardrail_denied'")
    message: str = Field(description="Human-readable explanation")
    detail: dict[str, Any] | None = None


PaymentStatusFilter = Literal["all", "failed", "captured", "created", "authorized", "refunded"]

File: frontend/src/lib/schemas.ts
/**
 * Runtime validation for the API boundary.
 *
 * `src/lib/types.ts` mirrors the Python schemas by hand, which is a deliberate
 * choice (see docs/08-DESIGN-DECISIONS.md, D16) but comes with an obvious risk:
 * a TypeScript interface is erased at build time, so if the backend renames a
 * field the compiler is perfectly happy and the value silently becomes
 * `undefined` somewhere deep in a component. On the approval screen, a field
 * that quietly becomes `undefined` is not a rendering glitch -- `can_approve`
 * arriving as `undefined` reads as falsy, and the operator is told a case cannot
 * be approved when policy says it can.
 *
 * These schemas are where that drift is caught: at the boundary, immediately,
 * with a message naming the field.
 *
 * **Why not validate every response.** Zod parsing costs bundle size and a little
 * time per call, and for most payloads a missing field is visibly cosmetic -- a
 * blank cell in a table that a reviewer would notice at a glance. The schemas
 * below cover the responses where silent drift would change a *decision* or
 * undermine a *claim*:
 *
 *   - `RecoveryCase`      the approval screen: amounts, guardrails, can_approve
 *   - `CheckoutSession`   the money screen: what the customer is asked to pay
 *   - `AuditChainVerification`  the integrity claim itself
 *   - `DashboardMetrics`  the headline numbers
 *   - `Policy` / `SystemStatus`  the limits, and the honesty badges
 *
 * Objects are non-strict on purpose: Zod strips unknown keys rather than
 * rejecting them, so the backend adding a field never breaks a deployed client.
 * The direction that matters is a field going *missing*, and that still fails.
 */

import { z } from "zod";

import {
  ACTOR_TYPE_LABEL,
  AGENT_MODE_LABEL,
  AUDIT_EVENT_LABEL,
  FAILURE_CATEGORY_LABEL,
  GATEWAY_MODE_LABEL,
  GUARDRAIL_DECISION_LABEL,
  PAYMENT_METHOD_LABEL,
  PAYMENT_STATUS_LABEL,
  RECOVERY_STATUS_LABEL,
  RECOVERY_STRATEGY_LABEL,
  TOOL_CAPABILITY_LABEL,
  type AuditChainVerification,
  type CheckoutSession,
  type DashboardMetrics,
  type Policy,
  type RecoveryCase,
  type SystemStatus,
} from "@/lib/types";

/**
 * Build a Zod enum from one of the display-label maps in `types.ts`.
 *
 * The label maps are typed `Record<Union, string>`, so their keys are exactly the
 * union members -- which makes them the single source of truth for both the
 * compile-time type and this runtime check. Writing the literals out again here
 * would create a third copy to keep in sync, and the copy that drifts is always
 * the one nobody is looking at.
 *
 * The cast is unavoidable: `z.enum` wants a non-empty tuple of literals and
 * `Object.keys` can only be known to return `string[]`. It is sound here because
 * the input is a `Record` over a string union, which cannot be empty in practice.
 */
function enumOfLabels<T extends string>(labels: Record<T, unknown>) {
  const values = Object.keys(labels) as [T, ...T[]];
  return z.enum(values);
}

const paymentStatus = enumOfLabels(PAYMENT_STATUS_LABEL);
const paymentMethod = enumOfLabels(PAYMENT_METHOD_LABEL);
const failureCategory = enumOfLabels(FAILURE_CATEGORY_LABEL);
const recoveryStrategy = enumOfLabels(RECOVERY_STRATEGY_LABEL);
const recoveryStatus = enumOfLabels(RECOVERY_STATUS_LABEL);
const guardrailDecision = enumOfLabels(GUARDRAIL_DECISION_LABEL);
const actorType = enumOfLabels(ACTOR_TYPE_LABEL);
const auditEventType = enumOfLabels(AUDIT_EVENT_LABEL);
const agentMode = enumOfLabels(AGENT_MODE_LABEL);
const gatewayMode = enumOfLabels(GATEWAY_MODE_LABEL);
const toolCapability = enumOfLabels(TOOL_CAPABILITY_LABEL);

/** Free-form JSON, for payloads and policy snapshots we render but never branch on. */
const jsonRecord = z.record(z.string(), z.unknown());

// ---------------------------------------------------------------------------
// Building blocks
// ---------------------------------------------------------------------------

export const customerSchema = z.object({
  id: z.string(),
  name: z.string(),
  email: z.string(),
  phone: z.string(),
  risk_flagged: z.boolean(),
  total_payments: z.number(),
  successful_payments: z.number(),
  prior_success_rate: z.number(),
  lifetime_value_paise: z.number(),
  lifetime_value_rupees: z.number(),
});

export const paymentSchema = z.object({
  id: z.string(),
  customer_id: z.string(),
  customer: customerSchema.nullable(),
  amount_paise: z.number(),
  amount_rupees: z.number(),
  currency: z.string(),
  method: paymentMethod,
  status: paymentStatus,
  description: z.string(),
  razorpay_order_id: z.string().nullable(),
  razorpay_payment_id: z.string().nullable(),
  error_code: z.string().nullable(),
  error_source: z.string().nullable(),
  error_step: z.string().nullable(),
  error_reason: z.string().nullable(),
  error_description: z.string().nullable(),
  is_recovery_attempt: z.boolean(),
  parent_payment_id: z.string().nullable(),
  recovery_case_id: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const guardrailEvaluationSchema = z.object({
  rule_id: z.string(),
  name: z.string(),
  description: z.string(),
  decision: guardrailDecision,
  passed: z.boolean(),
  reason: z.string(),
  observed: z.string().nullable(),
  limit: z.string().nullable(),
  // Defaulted rather than required: this field was added after the first
  // version of the checklist, and an older backend omitting it should degrade to
  // "every rule was consulted" -- the previous behaviour -- rather than blanking
  // the whole approval screen.
  applicable: z.boolean().default(true),
});

export const recoveryAttemptSchema = z.object({
  id: z.string(),
  attempt_number: z.number(),
  strategy: recoveryStrategy,
  amount_paise: z.number(),
  amount_rupees: z.number(),
  status: z.string(),
  gateway_mode: gatewayMode,
  razorpay_order_id: z.string().nullable(),
  razorpay_payment_id: z.string().nullable(),
  failure_reason: z.string().nullable(),
  created_at: z.string(),
  completed_at: z.string().nullable(),
});

export const agentToolCallSchema = z.object({
  id: z.string(),
  run_id: z.string(),
  step: z.number(),
  tool_name: z.string(),
  capability: toolCapability,
  arguments: jsonRecord,
  result: jsonRecord,
  ok: z.boolean(),
  error: z.string().nullable(),
  latency_ms: z.number(),
  created_at: z.string(),
});

export const auditEventSchema = z.object({
  id: z.string(),
  sequence: z.number(),
  case_id: z.string().nullable(),
  payment_id: z.string().nullable(),
  event_type: auditEventType,
  actor_type: actorType,
  actor_id: z.string(),
  summary: z.string(),
  payload: jsonRecord,
  prev_hash: z.string(),
  hash: z.string(),
  created_at: z.string(),
});

// ---------------------------------------------------------------------------
// Decision-critical responses
// ---------------------------------------------------------------------------

export const recoveryCaseSchema = z.object({
  id: z.string(),
  original_payment_id: z.string(),
  customer_id: z.string(),
  customer_name: z.string(),
  status: recoveryStatus,
  failure_category: failureCategory,
  strategy: recoveryStrategy,
  guardrail_decision: guardrailDecision,
  propensity_score: z.number(),
  agent_mode: agentMode,
  amount_paise: z.number(),
  amount_rupees: z.number(),
  attempt_count: z.number(),
  created_at: z.string(),
  updated_at: z.string(),

  classification_confidence: z.number(),
  agent_rationale: z.string(),
  customer_message: z.string(),
  propensity_model_version: z.string(),
  propensity_is_fallback: z.boolean(),

  guardrail_evaluations: z.array(guardrailEvaluationSchema),
  policy_snapshot: jsonRecord,

  approved_by: z.string().nullable(),
  approved_at: z.string().nullable(),
  rejected_by: z.string().nullable(),
  rejected_at: z.string().nullable(),
  rejection_reason: z.string().nullable(),

  recovered_at: z.string().nullable(),
  recovered_amount_paise: z.number(),
  recovered_amount_rupees: z.number(),
  failure_note: z.string().nullable(),
  expires_at: z.string().nullable(),

  original_payment: paymentSchema.nullable(),
  customer: customerSchema.nullable(),
  attempts: z.array(recoveryAttemptSchema),

  // The three fields that make this schema worth its weight. They are computed
  // server-side precisely so the client never re-implements policy; if any of
  // them went missing, `undefined` would read as `false` and the UI would refuse
  // an approval the policy engine had already allowed.
  can_approve: z.boolean(),
  can_reject: z.boolean(),
  approval_blocked_reason: z.string().nullable(),
}) satisfies z.ZodType<RecoveryCase, z.ZodTypeDef, unknown>;

export const checkoutSessionSchema = z.object({
  case_id: z.string(),
  attempt_id: z.string(),
  order_id: z.string(),
  amount_paise: z.number(),
  amount_rupees: z.number(),
  currency: z.string(),
  razorpay_key_id: z.string().nullable(),
  gateway_mode: gatewayMode,
  customer_name: z.string(),
  customer_email: z.string(),
  customer_phone: z.string(),
  description: z.string(),
  expires_at: z.string().nullable(),
}) satisfies z.ZodType<CheckoutSession, z.ZodTypeDef, unknown>;

export const auditChainVerificationSchema = z.object({
  valid: z.boolean(),
  events_checked: z.number(),
  head_hash: z.string().nullable(),
  broken_at_sequence: z.number().nullable(),
  message: z.string(),
}) satisfies z.ZodType<AuditChainVerification, z.ZodTypeDef, unknown>;

export const dashboardMetricsSchema = z.object({
  total_volume_paise: z.number(),
  captured_volume_paise: z.number(),
  failed_volume_paise: z.number(),
  recoverable_volume_paise: z.number(),
  recovered_volume_paise: z.number(),
  total_volume_rupees: z.number(),
  captured_volume_rupees: z.number(),
  failed_volume_rupees: z.number(),
  recoverable_volume_rupees: z.number(),
  recovered_volume_rupees: z.number(),
  total_payments: z.number(),
  failed_payments: z.number(),
  unanalysed_failures: z.number(),
  cases_total: z.number(),
  cases_awaiting_approval: z.number(),
  cases_blocked: z.number(),
  cases_recovered: z.number(),
  cases_failed: z.number(),
  recovery_rate_pct: z.number(),
  failure_rate_pct: z.number(),
  daily_budget_used_paise: z.number(),
  daily_budget_limit_paise: z.number(),
}) satisfies z.ZodType<DashboardMetrics, z.ZodTypeDef, unknown>;

export const policySchema = z.object({
  max_recovery_attempts: z.number(),
  recovery_cooldown_seconds: z.number(),
  high_value_review_threshold_paise: z.number(),
  max_recovery_amount_paise: z.number(),
  daily_recovery_budget_paise: z.number(),
  max_cases_per_customer_per_day: z.number(),
  min_propensity_score: z.number(),
  max_payment_age_hours: z.number(),
  require_human_approval: z.boolean(),
  auto_approve_enabled: z.boolean(),
  auto_approve_max_paise: z.number(),
  auto_approve_min_propensity: z.number(),
  recovery_link_ttl_minutes: z.number(),
  non_recoverable_categories: z.array(failureCategory),
  rules: z.array(
    z.object({
      rule_id: z.string(),
      name: z.string(),
      description: z.string(),
    }),
  ),
}) satisfies z.ZodType<Policy, z.ZodTypeDef, unknown>;

export const systemStatusSchema = z.object({
  app: z.string(),
  version: z.string(),
  environment: z.string(),
  agent_mode: agentMode,
  gemini_model: z.string().nullable(),
  gateway_mode: gatewayMode,
  ml_model_loaded: z.boolean(),
  ml_model_version: z.string().nullable(),
  database: z.string(),
  warnings: z.array(z.string()),
}) satisfies z.ZodType<SystemStatus, z.ZodTypeDef, unknown>;

/**
 * Validate a decoded response, turning a Zod failure into a message that names
 * the offending field.
 *
 * Thrown as a plain `Error` and then wrapped by `apiFetch` into an
 * `ApiRequestError`, so every consumer keeps handling exactly one error type.
 */
export function parseResponse<T>(
  schema: z.ZodType<T, z.ZodTypeDef, unknown>,
  data: unknown,
  label: string,
): T {
  const result = schema.safeParse(data);
  if (result.success) {
    return result.data;
  }
  const detail = result.error.issues
    .slice(0, 4)
    .map((issue) => `${issue.path.join(".") || "(root)"}: ${issue.message}`)
    .join("; ");
  throw new Error(
    `The API returned a ${label} that does not match the expected shape (${detail}). ` +
      "The frontend types and the backend schema have drifted - check " +
      "backend/app/domain/schemas.py against frontend/src/lib/types.ts.",
  );
}
